samtranslator/swagger/swagger.py

```diff
@@ -800,26 +803,206 @@ def add_resource_policy(self, resource_policy):
-            else:
-                statement = self.resource_policy['Statement']
-                if isinstance(statement, list):
-                    statement.extend(custom_statements)
-                else:
-                    statement = [statement]
-                    statement.extend(custom_statements)
+        self._doc[self._X_APIGW_POLICY] = self.resource_policy
+
+    def _add_iam_resource_policy_for_method(self, policy_list, effect, resource_list):
+        """
+        This method generates a policy statement to grant/deny specific IAM users access to the API method and
+        appends it to the swagger under `x-amazon-apigateway-policy`
+        :raises ValueError: If the effect passed in does not match the allowed values.
+        """
+        if not policy_list:
+            return
+
+        if effect not in ["Allow", "Deny"]:
+            raise ValueError('Effect must be one of {}'.format(['Allow', 'Deny']))
+
+        if not isinstance(policy_list, (dict, list)):
+            raise InvalidDocumentException(
+                [InvalidTemplateException("Type of '{}' must be a list or dictionary"
+                                          .format(policy_list))])
+
+        if not isinstance(policy_list, list):
+            policy_list = [policy_list]
+
+        self.resource_policy['Version'] = '2012-10-17'
+        policy_statement = {}
+        policy_statement['Effect'] = effect
+        policy_statement['Action'] = "execute-api:Invoke"
+        policy_statement['Resource'] = resource_list
+        policy_statement['Principal'] = {"AWS": policy_list}
+
+        if self.resource_policy.get('Statement') is None:
+            self.resource_policy['Statement'] = policy_statement
+        else:
+            statement = self.resource_policy['Statement']
+            if not isinstance(statement, list):
+                statement = [statement]
+            statement.extend([policy_statement])
+            self.resource_policy['Statement'] = statement
+
+    def _get_method_path_uri_list(self, path, api_id, stage):
+        """
+        It turns out that APIGW doesn't like trailing slashes in paths (#665)
+        and removes as a part of their behavior, but this isn't documented.
+        The regex removes the trailing slash to ensure the permission works as intended
+        """
+        methods = list(self.get_path(path).keys())
 
-                self.resource_policy['Statement'] = statement
+        uri_list = []
+        path = SwaggerEditor.get_path_without_trailing_slash(path)
 
-            self._doc[self._X_APIGW_POLICY] = self.resource_policy
+        for m in methods:
+            method = '*' if (m.lower() == self._X_ANY_METHOD or m.lower() == 'any') else m.upper()
+
+            # RestApiId can be a simple string or intrinsic function like !Ref. Using Fn::Sub will handle both cases
+            resource = '${__ApiId__}/' + '${__Stage__}/' + method + path
+            partition = ArnGenerator.get_partition_name(None)
```

Comment:
Do you think we should keep doing this (I know we do it elsewhere), or should we use ${AWS::Partition} pseudo parameter?

Reply:
I think we should change that in one big sweeping change in all tests.